Bottom: 2f70c8b131dad53212d3bfc722c503a571831b4b
Top:    0470bce436c7b350b0fc943f0ffcbb6dcbeda850
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2008-09-30 23:30:05 +0100

Refresh of double_click_opens_libs

---

diff --git a/src/hid/gtk/gui-library-window.c b/src/hid/gtk/gui-library-window.c
index 9e9cae1..c1f2ce7 100644
--- a/src/hid/gtk/gui-library-window.c
+++ b/src/hid/gtk/gui-library-window.c
@@ -265,6 +265,38 @@ lib_model_filter_visible_func (GtkTreeModel * model,
 }
 
 
+/*! \brief Handles activation (e.g. double-clicking) of a component row
+ *  \par Function Description
+ *  Component row activated handler:
+ *  As a convenince to the user, expand / contract any node with children.
+ *
+ *  \param [in] tree_view The component treeview.
+ *  \param [in] path      The GtkTreePath to the activated row.
+ *  \param [in] column    The GtkTreeViewColumn in which the activation occurre
+ *  \param [in] user_data The component selection dialog.
+ */
+static void
+tree_row_activated (GtkTreeView       *tree_view,
+                    GtkTreePath       *path,
+                    GtkTreeViewColumn *column,
+                    gpointer           user_data)
+{
+  GtkTreeModel *model;
+  GtkTreeIter iter;
+
+  model = gtk_tree_view_get_model (tree_view);
+  gtk_tree_model_get_iter (model, &iter, path);
+
+  if (!gtk_tree_model_iter_has_child (model, &iter))
+    return;
+
+  if (gtk_tree_view_row_expanded (tree_view, path))
+    gtk_tree_view_collapse_row (tree_view, path);
+  else
+    gtk_tree_view_expand_row (tree_view, path, FALSE);
+}
+
+
 /*! \brief Handles changes in the treeview selection.
  *  \par Function Description
  *  This is the callback function that is called every time the user
@@ -560,6 +592,12 @@ create_lib_treeview (GhidLibraryWindow * library_window)
 					  "model", model,
 					  "rules-hint", TRUE,
 					  "headers-visible", FALSE, NULL));
+
+  g_signal_connect (libtreeview,
+                    "row-activated",
+                    G_CALLBACK (tree_row_activated),
+                    compselect);
+
   /* connect callback to selection */
   selection = gtk_tree_view_get_selection (GTK_TREE_VIEW (libtreeview));
   gtk_tree_selection_set_mode (selection, GTK_SELECTION_SINGLE);
